Index: linux-2.6.32.48/drivers/pci/Kconfig
===================================================================
--- linux-2.6.32.48.orig/drivers/pci/Kconfig	2011-11-21 13:23:21.000000000 +0200
+++ linux-2.6.32.48/drivers/pci/Kconfig	2011-11-21 13:31:05.241666270 +0200
@@ -51,6 +51,12 @@
 
 	  When in doubt, say N.
 
+config PCI_DISABLE_COMMON_QUIRKS
+	bool "PCI disable common quirks"
+	depends on PCI
+	help
+	  If you don't know what to do here, say N.
+
 config HT_IRQ
 	bool "Interrupts on hypertransport devices"
 	default y
Index: linux-2.6.32.48/drivers/pci/quirks.c
===================================================================
--- linux-2.6.32.48.orig/drivers/pci/quirks.c	2011-11-21 13:23:21.000000000 +0200
+++ linux-2.6.32.48/drivers/pci/quirks.c	2011-11-21 13:34:35.919563819 +0200
@@ -96,6 +96,7 @@
 }
 DECLARE_PCI_FIXUP_HEADER(PCI_ANY_ID, PCI_ANY_ID, quirk_resource_alignment);
 
+#ifndef CONFIG_PCI_DISABLE_COMMON_QUIRKS
 /* The Mellanox Tavor device gives false positive parity errors
  * Mark this device with a broken_parity_status, to allow
  * PCI scanning code to "skip" this now blacklisted device.
@@ -1952,7 +1953,9 @@
 	}
 }
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_NCR, PCI_DEVICE_ID_NCR_53C810, fixup_rev1_53c810);
+#endif /* !CONFIG_PCI_DISABLE_COMMON_QUIRKS */
 
+#ifndef CONFIG_PCI_DISABLE_COMMON_QUIRKS
 /* Enable 1k I/O space granularity on the Intel P64H2 */
 static void __devinit quirk_p64h2_1k_io(struct pci_dev *dev)
 {
@@ -2549,6 +2552,7 @@
 	}
 }
 DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_TI, 0xb800, fixup_ti816x_class);
+#endif /* !CONFIG_PCI_DISABLE_COMMON_QUIRKS */
 
 static void pci_do_fixups(struct pci_dev *dev, struct pci_fixup *f,
 			  struct pci_fixup *end)
