--- a/include/asm-mips/system.h
+++ b/include/asm-mips/system.h
@@ -269,15 +269,15 @@ extern __u64 __xchg_u64_unsupported_on_3
 #define __xchg_u64 __xchg_u64_unsupported_on_32bit_kernels
 #endif
 
 /* This function doesn't exist, so you'll get a linker error
    if something tries to do an invalid xchg().  */
 extern void __xchg_called_with_bad_pointer(void);
 
-static inline unsigned long __xchg(unsigned long x, volatile void * ptr, int size)
+static __always_inline unsigned long __xchg(unsigned long x, volatile void * ptr, int size)
 {
 	switch (size) {
 		case 4:
 			return __xchg_u32(ptr, x);
 		case 8:
 			return __xchg_u64(ptr, x);
 	}
