--- a/arch/mips/mm/tlbex.c
+++ b/arch/mips/mm/tlbex.c
@@ -848,15 +848,14 @@ static __init void build_tlb_write_entry
 	case CPU_PR4450:
 		i_nop(p);
 		tlbw(p);
 		break;
 
 	case CPU_R10000:
 	case CPU_R12000:
-	case CPU_4KC:
 	case CPU_SB1:
 	case CPU_SB1A:
 	case CPU_4KSC:
 	case CPU_20KC:
 	case CPU_25KF:
 		tlbw(p);
 		break;
@@ -876,14 +875,15 @@ static __init void build_tlb_write_entry
 		i_nop(p);
 		i_nop(p);
 		i_nop(p);
 		i_nop(p);
 		tlbw(p);
 		break;
 
+	case CPU_4KC:
 	case CPU_4KEC:
 	case CPU_24K:
 	case CPU_34K:
 		i_ehb(p);
 		tlbw(p);
 		break;
 
