--- linux-2.6.15-5.2/net/ipv4/icmp.c
+++ linux-2.6.15.7/net/ipv4/icmp.c
@@ -524,7 +524,7 @@
 					  iph->tos;
 
 	if (ip_options_echo(&icmp_param.replyopts, skb_in))
-		goto ende;
+		goto out_unlock;
 
 
 	/*
--- linux-2.6.15-5.2/net/ipv4/ip_output.c
+++ linux-2.6.15.7/net/ipv4/ip_output.c
@@ -1237,11 +1237,7 @@
 	iph->tos = inet->tos;
 	iph->tot_len = htons(skb->len);
 	iph->frag_off = df;
-	if (!df) {
-		__ip_select_ident(iph, &rt->u.dst, 0);
-	} else {
-		iph->id = htons(inet->id++);
-	}
+	ip_select_ident(iph, &rt->u.dst, sk);
 	iph->ttl = ttl;
 	iph->protocol = sk->sk_protocol;
 	iph->saddr = rt->rt_src;

