--- a/drivers/net/phy/phy.c
+++ b/drivers/net/phy/phy.c
@@ -284,6 +284,7 @@ int phy_ethtool_sset(struct phy_device *
 
 	return 0;
 }
+EXPORT_SYMBOL(phy_ethtool_sset);
 
 int phy_ethtool_gset(struct phy_device *phydev, struct ethtool_cmd *cmd)
 {
@@ -300,7 +301,7 @@ int phy_ethtool_gset(struct phy_device *
 
 	return 0;
 }
-
+EXPORT_SYMBOL(phy_ethtool_gset);
 
 /* Note that this function is currently incompatible with the
  * PHYCONTROL layer.  It changes registers without regard to
@@ -355,6 +356,7 @@ int phy_mii_ioctl(struct phy_device *phy
 
 	return 0;
 }
+EXPORT_SYMBOL(phy_mii_ioctl);
 
 /* phy_start_aneg
  *
