Index: dnsmasq/src/dnsmasq.c
===================================================================
--- dnsmasq/src/dnsmasq.c
+++ dnsmasq/src/dnsmasq.c
@@ -221,16 +221,16 @@
 	  script_gid = ent_pw->pw_gid;
 	 }
-      else
-	baduser = daemon->scriptuser;
-    }
-  
-  if (daemon->username && !(ent_pw = getpwnam(daemon->username)))
-    baduser = daemon->username;
-  else if (daemon->groupname && !(gp = getgrnam(daemon->groupname)))
-    baduser = daemon->groupname;
-
-  if (baduser)
-    die(_("unknown user or group: %s"), baduser, EC_BADCONF);
-   
+//      else
+//	baduser = daemon->scriptuser;
+    }
+  
+//  if (daemon->username && !(ent_pw = getpwnam(daemon->username)))
+//    baduser = daemon->username;
+//  else if (daemon->groupname && !(gp = getgrnam(daemon->groupname)))
+//    baduser = daemon->groupname;
+
+//  if (baduser)
+//    die(_("unknown user or group: %s"), baduser, EC_BADCONF);
+
   /* implement group defaults, "dip" if available, or group associated with uid */
   if (!daemon->group_set && !gp)
