--- a/src/php.h
+++ b/src/php.h
@@ -263,20 +263,20 @@ extern request_rec *php_rqst;
  * most production web environments.
  */
 #define PHP_LOOPLIMIT 100000
 
 /*
  * Max size of a single line of input in the HTML files
  */
-#define LINEBUFSIZE	4096
+#define LINEBUFSIZE     512
 
 /*
  * Input buffer size for Exec() commands
  */
-#define EXEC_INPUT_BUF 4096
+#define EXEC_INPUT_BUF 512
 
 /*
  * ECHO_BUF sets the size of the echo overflow buffer
  * (only used in Apache module version)
  */
 #define ECHO_BUF 64
 
@@ -284,15 +284,15 @@ extern request_rec *php_rqst;
  * DEFAULT_MAX_DATA_SPACE
  *
  * Set this to the maximum size a memory sub-pool is allowed to grow
  * to.  (number of kilo bytes)  
  * In the Apache module version, this can be overwritten by the
  * phpMaxDataSpace Apache configuration directive.
  */
-#define DEFAULT_MAX_DATA_SPACE 4096
+#define DEFAULT_MAX_DATA_SPACE 1024
 
 /*
  * PHP_TRACK_VARS
  *
  * If you define this, 3 arrays will be created in your scripts.  They are:
  * 
  *  $PHP_GETVARS
