--- a/hostapd/config.c
+++ b/hostapd/config.c
@@ -190,14 +190,15 @@ static void hostapd_config_defaults_bss(
 #ifdef EAP_FAST
 	 /* both anonymous and authenticated provisioning */
 	bss->eap_fast_prov = 3;
 	bss->pac_key_lifetime = 7 * 24 * 60 * 60;
 	bss->pac_key_refresh_time = 1 * 24 * 60 * 60;
 #endif /* EAP_FAST */
         bss->macaddr_format = os_strdup(RADIUS_ADDR_FORMAT);
+        bss->mac_acl_empty_passwd = 0;
 }
 
 
 static struct hostapd_config * hostapd_config_defaults(void)
 {
 	struct hostapd_config *conf;
 	struct hostapd_bss_config *bss;
@@ -1528,14 +1529,16 @@ struct hostapd_config * hostapd_config_r
 			if (hostapd_config_read_macformat(pos, &bss->macaddr_format))
 			{
 				wpa_printf(MSG_ERROR, "Line %d: Failed to "
 					   "read macaddr_format '%s'",
 					   line, pos);
 				errors++;
 			}
+		} else if (os_strcmp(buf, "mac_acl_empty_passwd")== 0) {
+			bss->mac_acl_empty_passwd = atoi(pos);
 		} else if (os_strcmp(buf, "accept_mac_file") == 0) {
 			if (hostapd_config_read_maclist(pos, &bss->accept_mac,
 							&bss->num_accept_mac))
 			{
 				wpa_printf(MSG_ERROR, "Line %d: Failed to "
 					   "read accept_mac_file '%s'",
 					   line, pos);
--- a/hostapd/config.h
+++ b/hostapd/config.h
@@ -192,15 +192,16 @@ struct hostapd_bss_config {
 					* frames */
 
 	enum {
 		ACCEPT_UNLESS_DENIED = 0,
 		DENY_UNLESS_ACCEPTED = 1,
 		USE_EXTERNAL_RADIUS_AUTH = 2
 	} macaddr_acl;
-	char *macaddr_format; /* format patern for MAC address, kind of %2X:%2X:%2X:%2X:%2X:%2X or %2x%2x%2x%2x%2x%2x */
+        char *macaddr_format; /* format patern for MAC address, kind of %2X:%2X:%2X:%2X:%2X:%2X or %2x%2x%2x%2x%2x%2x */
+        int mac_acl_empty_passwd; /* no password for MAC authentication */
 	struct mac_acl_entry *accept_mac;
 	int num_accept_mac;
 	struct mac_acl_entry *deny_mac;
 	int num_deny_mac;
 
 	int auth_algs; /* bitfield of allowed IEEE 802.11 authentication
 			* algorithms, WPA_AUTH_ALG_{OPEN,SHARED,LEAP} */
--- a/hostapd/ieee802_11_auth.c
+++ b/hostapd/ieee802_11_auth.c
@@ -126,14 +126,16 @@ static int hostapd_radius_acl_query(stru
 	os_snprintf(buf, sizeof(buf), hapd->conf->macaddr_format, MAC2STR(addr));
 	if (!radius_msg_add_attr(msg, RADIUS_ATTR_USER_NAME, (u8 *) buf,
 				 os_strlen(buf))) {
 		wpa_printf(MSG_DEBUG, "Could not add User-Name");
 		goto fail;
 	}
 
+        if (hapd->conf->mac_acl_empty_passwd)
+                *buf = 0;
 	if (!radius_msg_add_attr_user_password(
 		    msg, (u8 *) buf, os_strlen(buf),
 		    hapd->conf->radius->auth_server->shared_secret,
 		    hapd->conf->radius->auth_server->shared_secret_len)) {
 		wpa_printf(MSG_DEBUG, "Could not add User-Password");
 		goto fail;
 	}
