--- a/agent/Makefile
+++ b/agent/Makefile
@@ -28,7 +28,7 @@
 #  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 #  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-LDFLAGS = -rdynamic -Wl,-export-dynamic
+LDFLAGS += -rdynamic -Wl,-export-dynamic
 LDLIBS = -ldebug -labz -lber -levent
 
 ifeq ($(shell uname -s),Linux)
--- a/tools/snmp/Makefile
+++ b/tools/snmp/Makefile
@@ -28,7 +28,7 @@
 #  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 #  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-LDFLAGS = -L$(TOPDIR)/manager
+LDFLAGS += -L$(TOPDIR)/manager
 LDLIBS = -ldebug -labz -lber -lmanager
 
 # path to toplevel directory from here
--- a/modules/ups/Makefile
+++ b/modules/ups/Makefile
@@ -30,7 +30,7 @@
 
 VERSION = $(shell pwd | sed 's|^.*/tinysnmp-\([^/]*\)/.*$$|\1|' | cut -f1 -d.)
 
-CPPFLAGS = -DVERSION=\"$(VERSION)\" 
+CPPFLAGS += -DVERSION=\"$(VERSION)\" 
 
 # path to toplevel directory from here
 TOPDIR = ../..
