--- a/src/drivers/driver_wext.c
+++ b/src/drivers/driver_wext.c
@@ -1938,18 +1938,25 @@ static void wpa_driver_wext_disconnect(s
 		/*
 		 * Clear the BSSID selection and set a random SSID to make sure
 		 * the driver will not be trying to associate with something
 		 * even if it does not understand SIOCSIWMLME commands (or
 		 * tries to associate automatically after deauth/disassoc).
 		 */
 		wpa_driver_wext_set_bssid(drv, null_bssid);
-
+		
+		/**
+		 * Do not cleanup/reset ssid
+		 * atheros driver goes mad because of that and tries associng
+		 * to wring essid, and when correct ssid is set -
+		 * it does not always update it correctly
 		for (i = 0; i < 32; i++)
 			ssid[i] = rand() & 0xFF;
 		wpa_driver_wext_set_ssid(drv, ssid, 32);
+		 *
+		 **/
 	}
 }
 
 
 static int wpa_driver_wext_deauthenticate(void *priv, const u8 *addr,
 					  int reason_code)
 {
