Index: linux-2.6.32.29/drivers/char/tty_io.c
===================================================================
--- linux-2.6.32.29.orig/drivers/char/tty_io.c
+++ linux-2.6.32.29/drivers/char/tty_io.c
@@ -1742,7 +1742,7 @@ retry_open:
 		extern struct tty_driver *console_driver;
 		driver = tty_driver_kref_get(console_driver);
 		index = fg_console;
-		noctty = 1;
+		/* noctty = 1; */ /* UBNT */
 		goto got_driver;
 	}
 #endif
@@ -1753,7 +1753,7 @@ retry_open:
 			if (driver) {
 				/* Don't let /dev/console block */
 				filp->f_flags |= O_NONBLOCK;
-				noctty = 1;
+				/* noctty = 1; */ /* UBNT */
 				goto got_driver;
 			}
 		}
