--- linux-2.6.31/arch/mips/kernel/module.c	2009-09-10 01:13:59.000000000 +0300
+++ mips-linux-2.6.31/arch/mips/kernel/module.c	2011-03-17 17:03:06.000000000 +0200
@@ -60,14 +60,30 @@
 #else
 	if (size == 0)
 		return NULL;
+#ifndef CONFIG_INSMOD_KSEG0
 	return vmalloc(size);
+#else
+	return (void *)alloc_pages_exact((size_t) size, GFP_KERNEL);
+#endif
 #endif
 }
 
 /* Free memory returned from module_alloc */
 void module_free(struct module *mod, void *module_region)
 {
+#ifndef CONFIG_INSMOD_KSEG0
 	vfree(module_region);
+#else
+	unsigned long size;
+	if(mod->module_core == module_region) {
+		size = (unsigned long)mod->core_size;
+		free_pages_exact((void *)module_region, (size_t)size);
+	}
+	else if(mod->module_init == module_region) {
+		size = (unsigned long)mod->init_size;
+		free_pages_exact((void *)module_region, (size_t)size);
+	}
+#endif
 }
 
 int module_frob_arch_sections(Elf_Ehdr *hdr, Elf_Shdr *sechdrs,
