--- linux-2.6.31/net/ipv4/netfilter/Kconfig	2009-09-10 01:13:59.000000000 +0300
+++ mips-linux-2.6.31/net/ipv4/netfilter/Kconfig	2011-03-17 17:03:19.000000000 +0200
@@ -388,5 +407,20 @@
 
 endif # IP_NF_ARPTABLES
 
+config NF_NAT_RTSP
+	tristate
+	depends on IP_NF_IPTABLES && NF_CONNTRACK && NF_NAT
+	default NF_NAT && NF_CONNTRACK_RTSP
+
+config NF_CONNTRACK_RTSP
+	tristate 'RTSP protocol support'
+	depends on NF_CONNTRACK
+	help
+	  Support the RTSP protocol.  This allows UDP transports to be setup
+	  properly, including RTP and RDT.
+
+	  If you want to compile it as a module, say 'M' here and read
+	  Documentation/modules.txt.  If unsure, say 'Y'.
+
 endmenu
 
