--- a/radvd.c
+++ b/radvd.c
@@ -200,17 +200,14 @@ main(int argc, char *argv[])
 	if (sock < 0)
 		exit(1);
 
 	/* check that 'other' cannot write the file
          * for non-root, also that self/own group can't either
          */
 	if (check_conffile_perm(username, conf_file) < 0) {
-		if (get_debuglevel() == 0)
-			exit(1);
-		else
 			flog(LOG_WARNING, "Insecure file permissions, but continuing anyway");
 	}
 	
 	/* if we know how to do it, check whether forwarding is enabled */
 	if (check_ip6_forwarding()) {
 		if (get_debuglevel() == 0) {
 			flog(LOG_ERR, "IPv6 forwarding seems to be disabled, exiting");
