Index: linux-2.6.32.49/net/wireless/wext.c
===================================================================
--- linux-2.6.32.49.orig/net/wireless/wext.c	2011-12-02 12:26:06.613022979 +0200
+++ linux-2.6.32.49/net/wireless/wext.c	2011-12-02 13:04:36.073021644 +0200
@@ -872,7 +873,6 @@
 
 	err = handler(dev, info, (union iwreq_data *) iwp, extra);
 
-	iwp->length += essid_compat;
 
 	/* If we have something to return to the user */
 	if (!err && IW_IS_GET(cmd)) {
@@ -887,7 +887,10 @@
 			err = -EFAULT;
 			goto out;
		}
+		if (essid_compat) ((char*)iwp->pointer)[iwp->length] = 0;
+
 	}
+	iwp->length += essid_compat;
 
 	/* Generate an event to notify listeners of the change */
 	if ((descr->flags & IW_DESCR_FLAG_EVENT) && err == -EIWCOMMIT) {
